File: leetcode/medium/1277_countSquares.py
from typing import List

class Solution:
    def countSquares_1(self, matrix: List[List[int]]) -> int:
        res = 0
        def checkMatrix(y, x):
            nonlocal matrix, res
            steps = min(len(matrix) - y, len(matrix[0]) - x) + 1
            for step in range(1, steps):
                for i in range(y, y + step):
                    for j in range(x, x + step):
                        if matrix[i][j] == 0:
                            return
                res += 1
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                checkMatrix(i, j)

        return res

    def countSquares(self, matrix: List[List[int]]) -> int:
        res = 0
        memo = [[0]*len(matrix[0]) for i in range(len(matrix))]

        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                val = matrix[i][j]
                if val == 0:
                    continue
                x = memo[i][j-1] if j > 0 else 0
                y = memo[i-1][j] if i > 0 else 0
                d = memo[i-1][j-1] if i > 0 and j > 0 else 0
                minSqr = min([x, y, d])
                memo[i][j] = minSqr + 1
                res += memo[i][j]

        return res


def test ():
    params = [
        {
            'input': [
              [0,1,1,1],
              [1,1,1,1],
              [0,1,1,1]
            ],
            'output': 15,
        },
        {
            'input': [
              [1,0,1],
              [1,1,0],
              [1,1,0]
            ],
            'output': 7,
        },
    ]
    solution = Solution()

    for param in params:
        result = solution.countSquares(param['input'])

        print(
            'SUCCESS' if result == param['output'] else 'ERROR',
            'input', param['input'],
            'output', param['output'],
            'result', result,
            '\n',
        )


if __name__ == '__main__':
    test()

File: leetcode/medium/1277_countSquares.ts
function countSquares(matrix: number[][]): number {
    let res: number = 0;

    for (let i: number = 0; i < matrix.length; ++i) {
        for (let j: number = 0; j < matrix[0].length; ++j) {
            if (matrix[i][j] === 0) continue;

            const top: number = i > 0 ? matrix[i-1][j] : 0;
            const left: number = j > 0 ? matrix[i][j-1] : 0;
            const diagonal: number = (i > 0 && j > 0) ? matrix[i-1][j-1] : 0;

            matrix[i][j] = Math.min(top, left, diagonal) + 1;
            res += matrix[i][j];
        }
    }

    return res;
};


/*
[0,1,1,1],
[1,1,1,1],
[0,1,1,1]


0,1,1,1 -   3
1,1,2,2 -   1+1+(1+1)+(1+1)=6
0,1,2,3 -   1+(1+1)+(1+1+1)=6
6+6+3=15


[1,0,1],
[1,1,0],
[1,1,0]

[1,0,1], 2
[1,1,0], 2
[1,2,0]   1+2 = 7

*/

const test = () => {
    const params = [
        {
            input: {
                matrix: [
                    [0,1,1,1],
                    [1,1,1,1],
                    [0,1,1,1]
                ],
            },
            output: 15,
        },
        {
            input: {
                matrix: [
                    [1,0,1],
                    [1,1,0],
                    [1,1,0]
                ],
            },
            output: 7,
        },
    ];

    params.forEach(({input, output}) => {
        const { matrix } = input;
        const result = countSquares(matrix);

        console.log(
            JSON.stringify(result) === JSON.stringify(output) ? 'SUCCESS ' : 'ERROR ',
            'input', JSON.stringify(input),
            'output', output,
            'result', result,
        );
    });
};

test();